src/main/java/com/avispl/symphony/dal/avdevices/audionetworkinterface/qsc/undnemo/QSCUndnemoCommunicator.java

```diff
@@ -0,0 +1,805 @@
+import com.avispl.symphony.dal.avdevices.audionetworkinterface.qsc.undnemo.utils.QSCUndnemoMetric;
+import com.avispl.symphony.dal.avdevices.audionetworkinterface.qsc.undnemo.utils.QSCUndnemoUDPCommand;
+import com.avispl.symphony.dal.util.StringUtils;
+
+/**
+ * QSC Attero Tech unDNEMO Adapter
+ *
+ * Monitoring:
+ * <ul>
+ * 	<li>Online / Offline Status</li>
+ * 	<li>Version Info</li>
+ * 	<li>Channel Info (channels 1-64)</li>
+ * 	<li>Active Channel Index (1-64)</li>
+ * 	<li>Speaker Mute (On/Off)</li>
+ * 	<li>Volume (1-10)</li>
+ * 	<li>Button Brightness (0-10)</li>
+ * 	<li>Display Brightness (0-10)</li>
+ * </ul>
+ *
+ * Controlling:
+ * <ul>
+ * 	<li>Set Active Channel Index (1-64)</li>
+ * 	<li>Set Speaker Mute (On/Off)</li>
+ * 	<li>Set Volume (1-10)</li>
+ * 	<li>Set Button Brightness (0-10)</li>
+ * 	<li>Set Display Brightness (0-10)</li>
+ * 	</ul
+ *
+ * @author Duy Nguyen
+ * @version 1.0
+ * @since 1.0
+ */
+public class QSCUndnemoCommunicator extends UDPCommunicator implements Monitorable, Controller {
+
+	/**
+	 * Process that is running whenever {@link QSCUndnemoCommunicator#getMultipleStatistics()} is called to fetch all 64 channel information.
+	 * - The worker thread will be destroyed when all 64 channels are fetched successfully or {@link QSCUndnemoCommunicator#internalDestroy()} is called.
+	 *
+	 * @author Maksym.Rossiytsev, Duy Nguyen
+	 * @since 1.0.0
+	 */
+	class QSCChannelDataLoader implements Runnable {
+
+		private volatile boolean inProgress;
+
+		private List<Integer> listIndexes;
+
+		/**
+		 * QSCChannelDataLoader with args constructor
+		 *
+		 * @param indexes list of index
+		 */
+		public QSCChannelDataLoader(List<Integer> indexes) {
+			inProgress = true;
+			listIndexes = indexes;
+		}
+
+		@Override
+		public void run() {
+			mainloop:
+			while (inProgress) {
+				if (!inProgress) {
+					break mainloop;
+				}
+				try {
+					retrieveChannelInfo(listIndexes);
+				} catch (Exception e) {
+					String errorMessage = String.format("Channel Info Data Retrieval-Error: %s with cause: %s", e.getMessage(), e.getCause().getMessage());
+					channelErrorMessagesList.add(errorMessage);
+					logger.error(errorMessage);
+				}
+				if (logger.isDebugEnabled()) {
+					logger.debug("Finished collecting channel info statistics cycle at " + new Date());
+				}
+				// Finished collecting, stop the thread. It will be called again when getMultipleStatistics is called
+				stop();
+			}
+		}
+
+		/**
+		 * Triggers main loop to stop
+		 */
+		public void stop() {
+			inProgress = false;
+		}
+	}
+
+	/**
+	 * Executor that runs all the async operations, that {@link #channelDataLoader} is posting
+	 */
+	private static ExecutorService executorService;
+
+	/**
+	 * Runner service responsible for collecting data
+	 */
+	private QSCChannelDataLoader channelDataLoader;
+
+	/**
+	 * List of channel info
+	 */
+	private List<ChannelInfo> channelInfoList = Collections.synchronizedList(new ArrayList<>());
```

Comment:
I think It should be final. Because you do not need to re-assign it

Reply:
Updated.